File: Bibliotek/Serialzation/BibMARCOutputStream.m
//
//  BibMARCOutputStream.m
//  Bibliotek
//
//  Created by Steve Brunwasser on 4/25/20.
//  Copyright © 2020 Steve Brunwasser. All rights reserved.
//

#import "BibMARCOutputStream.h"
#import "BibMARCSerialization.h"
#import "BibSerializationError+Internal.h"
#import "Bibliotek+Internal.h"

@implementation BibMARCOutputStream {
    NSOutputStream *_outputStream;
    NSStreamStatus _streamStatus;
    NSError *_streamError;
}

- (instancetype)init {
    return [self initToMemory];
}

- (instancetype)initWithOutputStream:(NSOutputStream *)outputStream {
    if (self = [super init]) {
        _outputStream = outputStream;
        _streamStatus = [outputStream streamStatus];
        _streamError = [outputStream streamError];
    }
    return self;
}

- (void)dealloc {
    [_outputStream close];
}

- (NSStreamStatus)streamStatus {
    return (_streamStatus == NSStreamStatusError) ? _streamStatus : [_outputStream streamStatus];
}

- (NSError *)streamError {
    return (_streamStatus == NSStreamStatusError) ? (_streamError ?: [_outputStream streamError]) : nil;
}

- (BOOL)hasSpaceAvailable {
    return [_outputStream hasSpaceAvailable];
}

- (NSData *)data {
    return [_outputStream propertyForKey:NSStreamDataWrittenToMemoryStreamKey];
}

- (instancetype)open {
    if (_streamStatus == NSStreamStatusNotOpen) {
        [_outputStream open];
        _streamStatus = [_outputStream streamStatus];
        _streamError = [_outputStream streamError];
    }
    return self;
}

- (instancetype)close {
    if (_streamStatus != NSStreamStatusClosed) {
        [_outputStream close];
        _streamStatus = [_outputStream streamStatus];
        _streamError = [_outputStream streamError];
    }
    return self;
}

- (BOOL)writeRecord:(BibRecord *)record error:(out NSError * _Nullable __autoreleasing *)error {
    switch ([self streamStatus]) {
        case NSStreamStatusOpen:
            break;
        case NSStreamStatusError:
            if (error != NULL) {
                *error = [self streamError];
            }
            return NO;
        default:
            if (error != NULL) {
                *error = BibSerializationMakeOutputStreamNotOpenedError(_outputStream);
            }
            return NO;
    }
    NSError *err = nil;
    BOOL const success = [BibMARCSerialization writeRecord:record toStream:_outputStream error:&err];
    if (!success) {
        _streamStatus = NSStreamStatusError;
        _streamError = err;
        if (error != NULL) {
            *error = err;
        }
    }
    return success;
}

@end

File: Bibliotek/Serialzation/BibMARCOutputStream.swift
//
//  BibMARCOutputStream.swift
//  Bibliotek
//
//  Created by Steve Brunwasser on 4/25/20.
//  Copyright © 2020 Steve Brunwasser. All rights reserved.
//

import Foundation

extension MARCOutputStream {
    /// Write a `Record` as MARC 21 data to the output stream.
    ///
    /// - parameter record: The record to write to the output stream.
    /// - returns: `true` when the record was successfully written to the output stream. Otherwise, `no` is returned.
    /// - precondition: The output stream must be opened.
    /// - postcondition: If an error is thrown when the output stream is opened, `streamStatus` and `streamError` will
    ///                  be set `NSStreamStatusError` and the thrown error respectively.
    /// - note: An error is thrown if the output stream has not yet been opened, but `streamStatus` and `streamError`
    ///         will not be set to reflect that failure. Attempting to write to the output stream after it is opened
    ///         will continue to write data as though it did not previously encounter an error.
    public func write(record: Record) throws -> Bool {
        return try self.__write(record as BibRecord)
    }
}
